README.md
[![Codecov test coverage](https://codecov.io/gh/eamena-project/eamenar/branch/main/graph/badge.svg)](https://app.codecov.io/gh/eamena-project/eamenar?branch=main)
# ***eamenaR*** <img src="https://raw.githubusercontent.com/eamena-project/eamena-arches-dev/main/www/eamenaR_logo.png" width='100px' align="right"/>
> R package for front-end statistical analysis of the EAMENA database

The ***eamenaR*** package allows to analyse the [typological](https://github.com/eamena-project/eamenaR#typology), [spatial](https://github.com/eamena-project/eamenaR#time) and [temporal](https://github.com/eamena-project/eamenaR#time) data, to [manage data](https://github.com/eamena-project/eamenaR#data-management), and to calculate basic statistics ([users](https://github.com/eamena-project/eamenaR#users), etc.) from the Arches-powered [EAMENA database](https://database.eamena.org/en/)[^8]. ***eamenaR*** is also open to new collaboration[^9]. 
  
```mermaid
flowchart LR
    A[(EAMENA<br>DB)] <--data<br>exchange--> B{{"eamenaR"}}:::eamenaRpkg;
    B --data<br>management--> B;
    B <--data<br>exchange--> C((third part<br>app));
    B --"output"--> D[maps<br>charts<br>listings<br>...];
    classDef eamenaRpkg fill:#e3c071;
```

The two main sources of data are: 
* GeoJSON files exported by [EAMENA searches](https://github.com/eamena-project/eamena-arches-dev/tree/main/data/geojson#readme); 
* data exported *via* a direct connection to the EAMENA PostgreSQL database (*restricted access*); 

The two main types of output are:
* static graphs and maps, for publication on paper;
* interactive graphs and maps for publication on the web (with Plotly[^6] and Leaflet[^7]);

Together with analysis functions, the package offers different methods to manage inputs and outputs from/to EAMENA. The ***eamenaR*** package makes the EAMENA DB data FAIR (*Findable, Accessible, Interoperable, Reusable*)

```mermaid
flowchart LR
    subgraph ide1 [<b>F</b>indable, <b>A</b>ccessible]
    A[(EAMENA<br>DB)];
    end
    A[(EAMENA<br>DB)] <---> B{{"eamenaR"}}:::eamenaRpkg;
    subgraph ide2 [<b>I</b>nteroperable, <b>R</b>eusable]
    B;
    end
    classDef eamenaRpkg fill:#e3c071;
```

## Families of functions

   
The functions names refer to their content :

| Function prefix      | Description                                            | Example                  |
| -----------          | -----------                                            |------                    |
| geojson_*            | all functions that deal with GeoJSON                   | `geojson_map()`          |
| geom_*               | any other function that deals with geometries          | `geom_bbox()`            |
| list_*               | structure a dataset                                    | `list_mapping_bu()`      |
| plot_*               | creates a map, a graphic, etc.                         | `plot_edtf()`            |
| ref_*                | direct connection to the EAMENA PostgreSQL database    | `ref_cultural_periods()` |

## Correspondances between concept labels and UUIDs

The [`ids.csv` file](https://github.com/eamena-project/eamenaR/blob/main/inst/extdata/ids.csv) is a correspondence table between permanent concepts' labels used in this package (`r.concept.name`), customised concepts' labels used in a specific Arches project `db.concept.name` and the latter UUIDs `db.concept.uuid` (by default, these values are those of the EAMENA DB). Depending on how you named your Arches instance concepts, you will have to modifiy these correspondences (see the function [`ref_ids()`](https://eamena-project.github.io/eamenaR/doc/ref_ids))

|r.concept.name            |db.concept.name           |db.concept.uuid                     |
|--------------------------|--------------------------|------------------------------------|
|id                        |EAMENA ID                 |34cfe992-c2c0-11ea-9026-02e7594ce0a0|
|Investigator.Role.Type    |Investigator Role Type    |d2e1ab96-cc05-11ea-a292-02e7594ce0a0|
|Geometric.Place.Expression|Geometric Place Expression|5348cf67-c2c5-11ea-9026-02e7594ce0a0|
|Cultural.Period           |Cultural Period           |3b5c9ac7-5615-3de6-9e2d-4cd7ef7460e4|
|Cultural.Sub-Period       |Cultural Sub-Period       |16cb160e-7b31-4872-b2ca-6305ad311011|
|Disturbance.Extent.Type   |Disturbance Extent Type   |41488800-6c00-30f2-b93f-785e38ab6251|


# Install and load package

Install the R package

```
devtools::install_github("eamena-project/eamenaR")
```

And load the package

```
library(eamenaR)
```

---

**How it works ?**  

The root directory on your local computer will be (*run*): `system.file(package = "eamenaR")`. By default, output will be saved in the `results/` folder. You can change this output folder by changing the `dirOut` option in the various functions to your choice. The `inst/extdata/` folder collects different sample files (GeoJSON, KML/KMZ, XLSX, etc.). 

---

# Data

Data will come from two sources: exported files, and SQL queries.

## Exported files

GeoJSON is the preferred format for working with EAMENA. Create a search in EAMENA, in the export menu, copy the **geojson** URL (in green) to the clipboard, paste it into your web browser and create a GeoJSON file[^1].  

![](https://raw.githubusercontent.com/eamena-project/eamena-arches-dev/main/www/geojson-export.png)

Paste the copied URL into your web browser and create a GeoJSON file, the result is something like :

![](https://raw.githubusercontent.com/eamena-project/eamena-arches-dev/main/www/geojson-url.png)

You can reformat the (Geo)JSON layout to make it more readable using https://codebeautify.org/jsonviewer. Copy the text content and save it in a new GeoJSON file, for example **caravanserail.geojson** ([rendered](https://github.com/eamena-project/eamena-arches-dev/blob/main/data/geojson/caravanserail.geojson) | [raw](https://raw.githubusercontent.com/eamena-project/eamena-arches-dev/main/data/geojson/caravanserail.geojson)).

## SQL queries

The PostgreSQL database is queried directly with SQL command passed through a DBI connection (`RPostgres::Postgres` driver).
# Typology

Whether the data is Heritage Places, Built Components, etc.

# Basic statistics

## Pie charts

The [`geojson_stat()`](https://eamena-project.github.io/eamenaR/doc/geojson_stat) function allows to display basic statistics. For example, a pie chart on 'Overall Condition Assessment':

```
geojson_stat(stat.name = "overall_cond",
             stat = "stats",
             field.names = c("Overall Condition State Type"),
             write.stat = T)
```
<p align="center">
  <img alt="img-name" src="./results/overall_condition_pie.png" width="400">
</p>

## Histograms

or an histogram on 'Disturbance Cause Type'

```
geojson_stat(stat.name = "distrub",
            stat = "stats",
            chart.type = "hist",
            field.names = c("Disturbance Cause Type"),
            fig.width = 10,
            fig.height = 9,
            write.stat = T)
```

<p align="center">
  <img alt="img-name" src="./results/distrub_cause_category_type_hist.png" width="600">
</p>

## Radar chart

or a radar chart on 'Resource Orientation'

```
geojson_stat(stat.name = "orientations",
             stat = "stats",
             chart.type = "radar",
             field.names = c("Resource Orientation"),
             fig.width = 9,
             fig.height = 8,
             write.stat = T)
```

<p align="center">
  <img alt="img-name" src="./results/orientations_radar.png" width="500">
</p>

## Boxplots

The [`geojson_boxplot()`](https://eamena-project.github.io/eamenaR/doc/geojson_boxplot) function creates boxplots. Path lengths, or areas, can be visualized in a boxplot, stratified by a variable (like "route") or not. With areas (`stat = area`, by default), each dot represents an heritage place. With path lenghts (`stat = dist`), each dot represent a segment length between two neighbouring caravanserails.

```
geojson_boxplot(stat = "area")
geojson_boxplot(stat = "dist")
```

<p align="center">
  <img alt="img-name" src="./results/caravanserais_area.png" width="400">
  <img alt="img-name" src="./results/caravanserais_dist.png" width="400">
</p>

Startified by routes and exported:

```
geojson_boxplot(stat = "area", by = "route", export.plot = T)
geojson_boxplot(stat = "dist", by = "route", export.plot = T)
```
<p align="center">
  <img alt="img-name" src="./results/caravanserais_areas_routes.png" width="700">
  <img alt="img-name" src="./results/caravanserais_dists_routes.png" width="700">
</p>

In the same way, these boxplot can be made interactive using Plotly, and exported as HTML files

```
geojson_boxplot(stat.name = "caravanserais_areas", stat = "area", by = "route",
                interactive = T,
                export.plot = T)
geojson_boxplot(stat.name = "caravanserais_dist", stat = "dist", by = "route",
                interactive = T,
                export.plot = T)
```

See these HTML files, [areas](https://eamena-project.github.io/eamenaR/results/caravanserais_areas) and [distances](https://eamena-project.github.io/eamenaR/results/caravanserais_dist)

# Spatial

Distribution maps for Heritages places and Geoarchaeology.

---

The [`ref_hps()`](https://eamena-project.github.io/eamenaR/doc/ref_hps) function allows a back-end connection.

## Heritages places

Using the default **caravanserail.geojson** ([rendered](https://github.com/eamena-project/eamena-arches-dev/blob/main/data/geojson/caravanserail.geojson) | [raw](https://raw.githubusercontent.com/eamena-project/eamena-arches-dev/main/data/geojson/caravanserail.geojson)) Heritage Places GeoJSON file

```
geojson_map(map.name = "caravanserail", fig.width = 11, export.plot = T)
```

<p align="center">
  <img alt="img-name" src="./results/caravanserail.png" width="700">
</p>

Maps can also be calculated on the values of GeoJSON fields, by adding the field names in the [`geojson_map()`](https://eamena-project.github.io/eamenaR/doc/geojson_map) function options.

```
geojson_map(map.name = "caravanserail",
            field.names = c("Damage Extent Type"),
            fig.width = 11,
            export.plot = T)
```

<p align="center">
  <img alt="img-name" src="./results/caravanserail_Damage Extent Type.png" width="700">
</p>

The color of the value (optional) is recorded in the [`symbology.xlsx`](https://github.com/eamena-project/eamenaR/blob/main/inst/extdata/symbology.xlsx) file  

<p align="center">
  <img alt="img-name" src="https://raw.githubusercontent.com/eamena-project/eamena-arches-dev/main/www/symbology-map.png" width="600">
  <br>
    <em>screenshot of the `symbology.xlsx` file registering the different colors of the values (only the columns `list`, `values` and `colors` are used)</em>
</p>

```
geojson_map(map.name = "caravanserail",
            field.names = c("Disturbance Cause Type ", "Damage Extent Type"),
            fig.width = 11,
            export.plot = T)
```

It will create two series of maps, one for each field (`"Disturbance Cause Type "`, `"Damage Extent Type"`) and because in `"Damage Extent Type"` there are multiple values for a same row, it creates as many maps as there are different values, here is an example:

<p align="center">
  <img alt="img-name" src="./results/caravanserail_Disturbance Cause Type _Human_Movement_Trampling.png" width="400">
  <img alt="img-name" src="./results/caravanserail_Disturbance Cause Type _Ploughing.png" width="400">
</p>

Finally, Plotly can be used to create an interactive map:

```
geojson_map(map.name = "caravanserail",
            geojson.path = paste0(exdata, "caravanserail_polygon.geojson"),
            plotly.plot = T,
            export.plot = F)
```

Will plot [this map](https://eamena-project.github.io/eamena-arches-dev/data/geojson/maps/caravanserail)

### Heritages places IDs ➡️ EAMENA ID 

Retrieve the matches between these maps' IDs and the EAMENA IDs for heritage places by running the [`geojson_stat()`](https://eamena-project.github.io/eamenaR/doc/geojson_stat) function:

```
geojson_stat(stat.name = "caravanserail", stat = "list_ids", export.stat = T)
```

This will give the data frame [caravanserail_list_ids.tsv](https://github.com/eamena-project/eamenaR/blob/main/results/caravanserail_list_ids.tsv). If you want the maps' IDs listed (e.g. for a figure caption), run :

```
geojson_stat(stat.name = "caravanserail", stat = "list_ids", export.stat = F)
```

Will give:

```
1: EAMENA-0192223, 2: EAMENA-0192598, 3: EAMENA-0192599, [...], 153: EAMENA-0194775, 154: EAMENA-0194776, 155: EAMENA-0194777, 156: EAMENA-0194778
```
### Paths

Reading the GeoJSON file of the heritage places, and [the CSV file](https://github.com/eamena-project/eamenaR/blob/main/inst/extdata/caravanserail_paths.csv) registering the paths between these heritage places, identified by different routes (route 1, route 2, etc.). Map them using the [`geojson_map_path()`](https://eamena-project.github.io/eamenaR/doc/geojson_map_path) function

```
geojson_map_path(map.name = "caravanserail_paths", export.plot = T, fig.width = 11)
```

<p align="center">
  <img alt="img-name" src="./results/caravanserail_paths.png" width="700">
</p>

#### Interactive

A good way to control the paths, avoiding double edges, etc. is to run an interactive plot of these paths:

```
geojson_map_path(interactive = T,
                 export.plot = F)
```
Will plot these five routes (from `0` to `4`) into an interactive [VisNetwork](https://datastorm-open.github.io/visNetwork/) HTML widget, for example [route 1](https://eamena-project.github.io/eamenaR/results/paths_map_route_1)

<p align="center">
  <img alt="img-name" src="./results/paths_map_route_1.png" width="500">
</p>

#### Profiles

Heritages places can be drawn with their elevation, for each route, using two functions: [`geojson_addZ()`](https://eamena-project.github.io/eamenaR/doc/geojson_addZ) to add a their Z value using a geoserver API and the function [`geojson_map_path()`](https://eamena-project.github.io/eamenaR/doc/geojson_map_path) to create the routes profiles (`export.type = "profile"`)

```
df <- geojson_addZ()
geojson_map_path(geojson.path = "C:/Rprojects/eamenaR/inst/extdata/caravanserailZ.geojson",
                 export.type = "profile",
                 export.plot = T,
                 fig.height = 11,
                 fig.width = 18)
```

<p align="center">
  <img alt="img-name" src="./results/map_paths_profile.png" width="700">
</p>

The numbers of the HP are the same as the [previous map](https://github.com/eamena-project/eamenaR#heritages-places-ids-%EF%B8%8F-eamena-id)

### Shape analysis

The use of POLYGONES (or even LINES), such as [caravanserail_polygon.geojson](https://github.com/eamena-project/eamenaR/blob/main/inst/extdata/caravanserail_polygon.geojson) allows to compute shape analysis. For the latter, we use the Momocs functions integrated in the [iconr](https://github.com/zoometh/iconr) package.
```
library(Momocs)
library(iconr)

dataDir <- "C:/Rprojects/eamena-arches-dev/projects/caravanserail"
nodes <- conv_geojson_to_wkt(dataDir = dataDir)
conv_wkt_to_jpg(nodes = nodes,
                ids = "site",
                dataDir = dataDir,
                out.dir = "_out")
```

`conv_geojson_to_wkt()` and `conv_wkt_to_jpg()` convert from GeoJSON to JPG, passing through WKT, creating this kind of outputs:

<p align="center">
  <img alt="img-name" src="./results/caravanserail/EAMENA-0164899.jpg" width="120">
  <img alt="img-name" src="./results/caravanserail/EAMENA-0164904.jpg" width="120">
  <img alt="img-name" src="./results/caravanserail/EAMENA-0164905.jpg" width="120">
  <img alt="img-name" src="./results/caravanserail/EAMENA-0164906.jpg" width="120">
  <img alt="img-name" src="./results/caravanserail/EAMENA-0164943.jpg" width="120">
  <img alt="img-name" src="./results/caravanserail/EAMENA-0164999.jpg" width="120">
</p>

These JPGs are analysed through shape analysis comparisons (here, we limit the study to 50 caravanserails)

```
dist <- morph_nds_compar(nodes = nodes,
                         cex = .5,
                         lwd = .5,
                         colored = FALSE,
                         dataDir = dataDir,
                         out.dir = "_out")
```

The variable `dist` stores the distance matrix between each pairs of caravanserails. The `morph_nds_compar()` output plots are :

<p align="center">
  <img alt="img-name" src="./results/caravanserail_compar_panel.png" width="700">
  <br>panel
</p>
<p align="center">
  <img alt="img-name" src="./results/caravanserail_compar_stack.png" width="700">
  <br>stack
</p>
<p align="center">
  <img alt="img-name" src="./results/caravanserail_compar_PCA.png" width="700">
  <br>PCA
</p>
<p align="center">
  <img alt="img-name" src="./results/caravanserail_group_clust.png" width="700">
  <br>HCA
</p>


After the shape comparisons, a classification can be made with `morph_nds_group()`. The HCA shows that there are two main groups (or centres, `nb.centers = 2`). We can reuse this parameter for shape classification:

```
mbrshp <- morph_nds_group(nodes = nodes,
                          nb.centers = 2,
                          dataDir = dataDir,
                          out.dir = "_out")
```
It gives a Kmeans plot with 2 centers:

<p align="center">
  <img alt="img-name" src="./results/caravanserail_group_kmeans.png" width="700">
  <br>Kmeans with 2 centers
</p>

The variable `mbrshp` stores the membership of all caravanserais (here group 1 or group 2). It can be reused in the [`geojson_map()`](https://eamena-project.github.io/eamenaR/doc/geojson_map) function, for example, to locate the different forms of caravanserai.

## Related resources

Part of the information of the Heritage Places can be recorded in the Built Components (`COMPONENT-`), which are connected components of Heritage Places. For example, `COMPONENT-0000141`, `COMPONENT-0000143` and `COMPONENT-0000144` record respectively 30 Stables, 1 Courtyard and 28 Rooms for the caravanserail Maranjab (`EAMENA-164943`).

<p align="center">
  <img alt="img-name" src="https://raw.githubusercontent.com/eamena-project/eamena-arches-dev/main/www/relat-resource-ex1.png" width="450">
</p>

By default, in EAMENA, relationships between Heritage Places and Built Component are of the type `PX_is_related_to`, unlike relationships between Heritage Places and Persons (`L33_has_maker`) or between Heritage Places and Grid Squares (`P89_falls_within`).  

Functions [`list_related_resources()`](https://eamena-project.github.io/eamenaR/doc/list_related_resources) allows to retrieve this data for a given Heritage Place

```
d <- hash::hash()
my_con <- RPostgres::dbConnect(drv = RPostgres::Postgres(),
                               user = 'xxx',
                               password = 'xxx',
                               dbname = 'eamena',
                               host = 'ec2-54-155-109-226.eu-west-1.compute.amazonaws.com',
                               port = 5432)

df <- list_related_resources(db.con = my_con,
                             d = d,
                             relationshiptype = "PX_is_related_to",
                             id = "EAMENA-0164943",
                             disconn = FALSE)
df
```

Will give this `df` dataframe, with the keys (`id` and `uuid`) of the connected components:

|          hp.id|                              hp.uuid|             cc.id|                              cc.uuid|
|---------------|-------------------------------------|------------------|-------------------------------------|
| EAMENA-0164943| d4feb830-10c7-4d80-a19e-e608f424be4c| COMPONENT-0000141| 90400bb6-ff54-4afd-8183-65c67fa97448|
| EAMENA-0164943| d4feb830-10c7-4d80-a19e-e608f424be4c| COMPONENT-0000143| 0dab164a-6d3a-443c-954a-50d93efbff35|
| EAMENA-0164943| d4feb830-10c7-4d80-a19e-e608f424be4c| COMPONENT-0000144| 28af281c-e4b9-44ac-aa98-2608581b7540|

Where `hp` is the Heritage place, and `cc` the connected component(s). The function [`select_related_resources()`](https://eamena-project.github.io/eamenaR/doc/select_related_resources) allows to retrieve the values of a given variable. For example, to retrieve the total number of Rooms, use the calculated dataframe `df` listing the keys (see [`list_related_resources()`](https://eamena-project.github.io/eamenaR/doc/list_related_resources)) and modify the parameter `having`. By default the value will be read in the field `"Measurement Number"` (function parameter `measure`).

```
df.measures <- select_related_resources(db.con = my_con,
                                        having = "Room",
                                        df = df)
df.measures
```

Will give this `df.measures` dataframe:

|hp.id          |hp.uuid                              |cc.id             |cc.uuid                              |cc.type | cc.measure|
|:--------------|:------------------------------------|:-----------------|:------------------------------------|:-------|----------:|
|EAMENA-0164943 |d4feb830-10c7-4d80-a19e-e608f424be4c |COMPONENT-0000144 |28af281c-e4b9-44ac-aa98-2608581b7540 |Room    |         28|

To retrieve the Heritage places' information about Rooms and Stables, create a dataframe to store this data, and run a loop stament over Heritage Places and types of Built components:

```
hps <- c("EAMENA-0164943", "EAMENA-0164937", "EAMENA-0164905")
bcs <- c("Room", "Stable")

df.measures.all <- data.frame(hp.id = character(),
                              hp.uuid = character(),
                              cc.id = character(),
                              cc.uuid = character(),
                              cc.type = character(),
                              cc.measure = numeric())

for(ea in hps){
  df <- list_related_resources(db.con = my_con,
                               d = d,
                               id = ea,
                               disconn = F)
  for(have in bcs){
    df.measures <- select_related_resources(db.con = my_con,
                                            df = df,
                                            having = have,
                                            disconn = F)
    df.measures.all <- rbind(df.measures.all, df.measures)
  }
}
df.measures.all[ , c("hp.id", "cc.id", "cc.type", "cc.measure")]
```

Will give:

|hp.id          |cc.id             |cc.type | cc.measure|
|:--------------|:-----------------|:-------|----------:|
|EAMENA-0164943 |COMPONENT-0000144 |Room    |         28|
|EAMENA-0164943 |COMPONENT-0000141 |Stable  |         30|
|EAMENA-0164937 |COMPONENT-0000148 |Room    |         37|
|EAMENA-0164937 |COMPONENT-0000149 |Stable  |         60|
|EAMENA-0164905 |COMPONENT-0000145 |Room    |         24|
|EAMENA-0164905 |COMPONENT-0000146 |Stable  |         30|


## Geoarchaeology

For MaREA geoarchaeological data, with the [`geojson_map()`](https://eamena-project.github.io/eamenaR/doc/geojson_map) function:

```
geojson_map(map.name = "geoarch",
            ids = "GEOARCH.ID",
            stamen.zoom = 6,
            geojson.path = "C:/Rprojects/eamena-arches-dev/data/geojson/geoarchaeo.geojson",
            export.plot = F)
```

<p align="center">
  <img alt="img-name" src="./results/geoarchaeo.png" width="450">
</p>

# Time

Either for [cultural periods](https://github.com/eamena-project/eamenaR#cultural-periods) or [EDTF](https://github.com/eamena-project/eamenaR#edtf) formats

## Cultural Periods

### Cultural and Subcultural periods references

Use the [`ref_cultural_periods()`](https://eamena-project.github.io/eamenaR/doc/ref_cultural_periods) and [`list_child_concepts()`](https://eamena-project.github.io/eamenaR/doc/list_child_concepts) to retrieve the list of cultural periods and subperiods directly from the EAMENA DB. 

```
# create an hash dictionnary to store the cultural ans subcultural periods
d <- hash::hash()
# replace 'xxx' with the username and password
my_con <- RPostgres::dbConnect(drv = RPostgres::Postgres(),
                               user = 'postgres',
                               password = 'postgis',
                               dbname = 'eamena',
                               host = 'ec2-54-155-109-226.eu-west-1.compute.amazonaws.com',
                               port = 5432)
# get cultural periods and subcultural periods
d <- list_child_concepts(db.con = my_con, d = d, 
                         field = "cultural_periods", 
                         concept.name = 'Cultural Period',
                         disconn = F)
d <- ref_cultural_periods(db.con = my_con, d = d,
                          field = "cultural_periods",
                          disconn = F)
d <- list_child_concepts(db.con = my_con, d = d, 
                         field = "cultural_subperiods", 
                         concept.name = 'Cultural Sub-Period',
                         disconn = F)
d <- ref_cultural_periods(db.con = my_con, d = d,
                          field = "cultural_subperiods")
# export as TSV
df.periods <- rbind(d$cultural_periods, d$cultural_subperiods)
tout <- paste0("C:/Rprojects/eamena-arches-dev/projects/periodo/cultural_periods.tsv")
write.table(df.periods, tout, sep ="\t", row.names = F)
```

Gives [this TSV dataframe](https://github.com/eamena-project/eamenaR/blob/main/results/cultural_periods.tsv) with (sub)cultural periods names, *tpq* and *taq*

<p align="center">
  <img alt="img-name" src="https://raw.githubusercontent.com/eamena-project/eamena-arches-dev/main/www/time-cultural-periods.png" width="550">
    <br>
  <em>screenshot of the</em> `cultural_periods.tsv` <em>dataframe</em>
</p>

---
  
**How it works ?**

These two functions connects the EAMENA DB to parse the arborescence of periods (parents) and superiods (childs) concepts (a tree-like structure) to retrieve their names, their start date (*tpq*) and end date (*taq*).

<p align="center">
  <img alt="img-name" src="https://raw.githubusercontent.com/eamena-project/eamena-arches-dev/main/www/time-cultural-periods-rdm.png" width="550">
  <br>
  <em>screenshot of the Reference Data Manager, parent node Cultural Period</em>
</p>

These latters (start date and end date) are stored in the `scopeNote` of each cultural periods and subperiods

<p align="center">
  <img alt="img-name" src="https://raw.githubusercontent.com/eamena-project/eamena-arches-dev/main/www/time-cultural-periods-rdm-1.png" width="550">
  <br>
  <em>screenshot of the Reference Data Manager, child node Palaeolithic (Levant/Mesopotamia/Arabia)</em>
</p>
  
---

### Plot cultural periods from a GeoJSON file

Create a hash dictonnary named `d` to store all data

```
library(hash)

d <- hash()
```

Store all periods and sub-periods represented in the GeoJSON in the `d` dictonnary, and plot them by EAMENA ID using the [`list_cultural_periods()`](https://eamena-project.github.io/eamenaR/doc/list_cultural_periods) function

```
d <- list_cultural_periods(db = "geojson", 
                           d = d)
plot_cultural_periods(d = d, field = "periods", plot.type = "by.eamenaid", export.plot = T)
plot_cultural_periods(d = d, field = "subperiods", plot.type = "by.eamenaid", export.plot = T)
```
<p align="center">
  <img alt="img-name" src="./results/cultural_periods_byeamenaid.png" width="700">
<br><br>
and superiods
<br><br>
  <img alt="img-name" src="./results/cultural_subperiods_byeamenaid.png" width="700">
</p>

Here, the [`plot_cultural_periods()`](https://eamena-project.github.io/eamenaR/doc/plot_cultural_periods) function  will export two PNG charts for the default **caravanserail.geojson** ([rendered](https://github.com/eamena-project/eamena-arches-dev/blob/main/data/geojson/caravanserail.geojson) | [raw](https://raw.githubusercontent.com/eamena-project/eamena-arches-dev/main/data/geojson/caravanserail.geojson)) file. Periods and subperiods represented in a GeoJSON file can also be summed in a histogram

```
plot_cultural_periods(d = d, field = "subperiods", plot.type = "histogram", export.plot = T)
```
<p align="center">
  <img alt="img-name" src="./results/cultural_subperiods_histog.png" width="700">
</p>

## EDTF

Performs an aoristic analysis. By default, the function reads the sample data `disturbances_edtf.xlsx` and performs the analysis by days (year-month-day: ``ymd``). Two graphs are created, one adding up all the threats, and the other where each category of threat is individualised.

Run the [`plot_edtf()`](https://eamena-project.github.io/eamenaR/doc/plot_edtf) function with the default parameters.

```
library(dplyr)

plot_edtf()
```

<p align="center">
  <img alt="img-name" src="./results/edtf_plotly_category_ymd.png" width="700">
</p>

Aggregate the dates by months (`"ym"`) by thearts categories.

```
plot_edtf(edtf_span = "ym", edtf_analyse = "category")
```

<p align="center">
  <img alt="img-name" src="./results/edtf_plotly_category_ym.png" width="700">
</p>

The interactive plotly output is [edtf_plotly_category_ym_threats_types.html](https://eamena-project.github.io/eamenaR/results/edtf_plotly_category_ym_threats_types.html)

# Users

The function [`ref_users()`](https://eamena-project.github.io/eamenaR/doc/ref_users) provides basic statistics on the users of the EAMENA database, for example by plotting the cumulative distribution function of the user first registration:

```
d <- hash::hash()
my_con <- RPostgres::dbConnect(drv = RPostgres::Postgres(),
                               user = 'xxx',
                               password = 'xxx',
                               dbname = 'eamena',
                               host = 'ec2-54-155-109-226.eu-west-1.compute.amazonaws.com',
                               port = 5432)
d <- ref_users(db.con = my_con,
               d = d,
               date.after = "2020-08-01",
               plot.g = T,
               fig.width = 14)
```

Here we restrict the plot to dates after 2020-08-01 (option `date.after`). The option `plot.g = T` gives this plot:

<p align="center">
  <img alt="img-name" src="./results/users_date_joined.png" width="600">
</p>

The total number of users can also be restricted to an interval (options `date.after` and `date.before`), for example limiting the count to the year 2022: 

```
d <- ref_users(db.con = my_con,
               stat.name = "users_date_joined_2",
               d = d,
               date.after = "2022-01-01",
               date.before = "2022-12-01",
               plot.g = T,
               export.plot.g = T,
               fig.width = 14)

```

<p align="center">
  <img alt="img-name" src="./results/users_date_joined_2.png" width="600">
</p>


```
d$total_users
#   count
# 1   480
```

The other statistic calculated is the total number of users (minus those who have an account but have never logged in)


# Data management

## Duplicates

The function [`ref_are_duplicates()`](https://eamena-project.github.io/eamenaR/doc/ref_are_duplicates) identifies potential duplicates in a GeoJSON file, or directly in the EAMENA database. Using a fuzzy match between the values of a selection of fields, for two HPs identified by their ResourceID, this function creates a data frame with the match score (`dist` column) between each field:

```
d <- hash::hash()
d <- ref_are_duplicates(d = d,
                        export.table = T,
                        fileOut = "duplicates.csv")
```

Creates this kind of table:
  
| field                           | 563567f7-eef0-4683-9e88-5e4be2452f80 | fb0a2ef4-023f-4d13-b931-132799bb7a6c | dist |
|---------------------------------|--------------------------------------|--------------------------------------|------|
| EAMENA ID                       | EAMENA-0207209                       | EAMENA-0182057                       |  -   |
| Assessment.Investigator...Actor | Hamed Rahnama                        | Hamed Rahnama, Bijan Rouhani         | 0.18 |
| Assessment.Activity.Date        | 2021-05-25                           | 2022-08-21, 2022-08-30               | 0.32 |
| Resource.Name                   | Bedasht Caravanserai, ..., CVNS-IR   | CVNS-IR, Bedasht Caravanserai, ...   | 0.26 |
| geometry                        | c(55.05059, 36.42466)                | c(55.05059, 36.42466)                | 0    |

The `dist` shows that the geometries are exactly the same, and that there are slight differences in the other fields. The CSV output is here: https://github.com/eamena-project/eamenaR/blob/main/results/duplicates.csv

## BU

The Bulk upload procedure

### BU append

```mermaid
flowchart LR
    A[(EAMENA<br>DB)] --1. GeoJSON<br><b>POINT</b>--> C("geojson_kml()"):::eamenaRfunction;
    A --1. GeoJSON<br><b>POINT</b>--> F("geojson_shp()"):::eamenaRfunction;
    F --2. SHP--> G((GIS));
    G --3. create<br><b>POLYGON</b>--> G;
    G --4. SHP--> F;
    C --2. KML/KMZ--> B((Google<br>Earth));
    B --3. create<br><b>POLYGON</b>--> B;
    B --4. KML/KMZ--> C;
    C --5. GeoJSON<br><b>POLYGON</b>--> D("geojson_csv()"):::eamenaRfunction;
    F --5. GeoJSON<br><b>POLYGON</b>--> D:::eamenaRfunction;
    D --6. append<br>new geometries--> A;
    classDef eamenaRfunction fill:#e7deca;
```

#### Integrating Google Earth geometries

Most of the geometries in EAMENA are POINTS (`Geometry Type` = `Center Point`). The objective is to acquire new geometries, like POLYGONs, created in third part app, like Google Earth or a GIS, and to append them to already existing records in EAMENA.


```mermaid
flowchart LR

NEWS.md
# Version 0.1.0 (1 Jan 2023)
First release.

R/_temp.R

library(devtools)

use_testthat()
use_test("geojson_addZ")
use_coverage(type = c("codecov"))

library(covr)
codecov()

######################################

# work with GET() + cookies

http.req <- "https://database.eamena.org/en/api/search/export_results?paging-filter=1&tiles=true&format=geojson&precision=6&total=223&term-filter=%5B%7B%22context%22%3A%22%22%2C%22context_label%22%3A%22Heritage%20Place%20-%20Resource%20Name%22%2C%22id%22%3A1%2C%22text%22%3A%22CVNS-IR%22%2C%22type%22%3A%22term%22%2C%22value%22%3A%22CVNS-IR%22%2C%22inverted%22%3Afalse%7D%5D"
r <- httr::GET(http.req,
               config = list(csrftoken="tzkEIDM1WP1yxoqzIT6ozXZmlJlvQfrhFwT0LrzXEYn25klUpS5iuyN5AsDser2L"))
rr <- httr::content(r)


############################
d <- hash::hash()
my_con <- RPostgres::dbConnect(drv = RPostgres::Postgres(),
                               user = 'postgres',
                               password = 'postgis',
                               dbname = 'eamena',
                               host = 'ec2-54-155-109-226.eu-west-1.compute.amazonaws.com',
                               port = 5432)
ref_users(db.con = my_con,
          d = d,
          date.after = "2022-01-01",
          date.before = "2022-12-01",
          plot.g = T)



stat.name = "orientations_path"
stat.name = "boxplot_by_routes"
geojson.path = "C:/Rprojects/eamenaR/inst/extdata/caravanserail.geojson"
csv.path = "C:/Rprojects/eamenaR/inst/extdata/caravanserail_paths.csv"
ids = eamenaR::ref_ids("hp.id", ids.path = "C:/Rprojects/eamenaR/inst/extdata/ids.csv")
dirOut = "C:/Rprojects/eamenaR/results/"
stat = "stats"
chart.type = "radar"
field.names = c("Resource Orientation")
by = "route"
write.stat = F

paths <- read.csv(csv.path)
df1 <- paths[ , c("from", "route")]
names(df1) <- c("id", "route")
df2 <- paths[ , c("to", "route")]
names(df2) <- c("id", "route")
df <- rbind(df1, df2)
head(df)
df <- df[!duplicated(df), ]

hp_routes <- ref_routes(geojson.path = "C:/Rprojects/eamenaR/inst/extdata/caravanserail.geojson",
                        csv.path = "C:/Rprojects/eamenaR/inst/extdata/caravanserail_paths.csv",
                        by = "route")



geojson_stat(stat.name = stat.name,
             geojson.path = geojson.path,
             csv.path = csv.path,
             ids = ids,
             dirOut = dirOut,
             stat = stat,
             chart.type = chart.type,
             field.names = field.names,
             by = by,
             write.stat = write.stat)
